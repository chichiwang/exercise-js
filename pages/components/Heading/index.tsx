import React, { useState } from 'react';
import classNames from 'classnames';

import { HeadingLevel } from 'app/types';
import LinkIcon from 'pages/components/LinkIcon';

import styles from './styles.module.scss';

interface HeadingProps extends React.PropsWithChildren {
  center?: boolean,
  className?: string,
  id?: string,
  level: HeadingLevel,
}

function Heading({
  level,
  center = false,
  children,
  className,
  id,
}: HeadingProps) {
  const [hovered, setHovered] = useState(false);

  function onMouseEnter() {
    setHovered(true);
  }

  function onMouseLeave() {
    setHovered(false);
  }

  const containerProps = {
    className: classNames(
      styles.container,
      {
        [styles.center]: center,
        [styles.hover]: hovered,
      },
    ),
    ...(typeof id === 'undefined' ? {} : {
      onMouseEnter,
      onMouseLeave,
    }),
  };

  return (
    <div {...containerProps}>
      <div className={styles.wrapper}>
        {
          React.createElement(
            `h${level}`,
            {
              ...(typeof id === 'undefined' ? {} : { id }),
              className: classNames(
                styles.tag,
                className,
              ),
            },
            children,
          )
        }
        {
          typeof id === 'undefined'
            ? null
            : (
              <a className={styles.link} href={`#${id}`} aria-labelledby={id}>
                <LinkIcon width="1.75rem" height="1.75rem" />
              </a>
            )
        }
      </div>
    </div>
  );
}

export default Heading;
