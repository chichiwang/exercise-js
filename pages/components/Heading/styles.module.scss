.container {
  display: flex;
  flex-direction: row;
  justify-content: left;
  
  .wrapper {
    position: relative;
    display: flex;
    align-items: center;

    .tag {
      display: inline-block;
    }

    .link {
      position: absolute;
      right: -30px;
    }
  }

  &.center {
    justify-content: center;
  }
}
