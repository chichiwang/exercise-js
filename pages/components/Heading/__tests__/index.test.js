import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Heading from '../index';

describe('Heading component', () => {
  test('matches uncentered snapshot', () => {
    const tree = renderer.create(
      <Heading level="2">
        H2 Heading
      </Heading>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('matches className snapshot', () => {
    const tree = renderer.create(
      <Heading level="4" className="foobar">
        H1 Heading
      </Heading>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders child components', () => {
    const childText = 'Paragraph Contents';

    render(
      <Heading level="3">
        <p>{childText}</p>
      </Heading>,
    );

    expect(screen.getByText(childText)).toBeInTheDocument();
  });

  describe('heading levels', () => {
    test('contains a h1 tag', () => {
      const headingText = 'H1 tag contents';

      render(
        <Heading level="1">
          {headingText}
        </Heading>,
      );

      expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent(headingText);
    });

    test('contains a h2 tag', () => {
      const headingText = 'H2 tag contents';

      render(
        <Heading level="2">
          {headingText}
        </Heading>,
      );

      expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(headingText);
    });

    test('contains a h3 tag', () => {
      const headingText = 'H3 tag contents';

      render(
        <Heading level="3">
          {headingText}
        </Heading>,
      );

      expect(screen.getByRole('heading', { level: 3 })).toHaveTextContent(headingText);
    });

    test('contains a h4 tag', () => {
      const headingText = 'H4 tag contents';

      render(
        <Heading level="4">
          {headingText}
        </Heading>,
      );

      expect(screen.getByRole('heading', { level: 4 })).toHaveTextContent(headingText);
    });

    test('contains a h5 tag', () => {
      const headingText = 'H4 tag contents';

      render(
        <Heading level="5">
          {headingText}
        </Heading>,
      );

      expect(screen.getByRole('heading', { level: 5 })).toHaveTextContent(headingText);
    });

    test('contains a h6 tag', () => {
      const headingText = 'H6 tag contents';

      render(
        <Heading level="6">
          {headingText}
        </Heading>,
      );

      expect(screen.getByRole('heading', { level: 6 })).toHaveTextContent(headingText);
    });
  });
});
