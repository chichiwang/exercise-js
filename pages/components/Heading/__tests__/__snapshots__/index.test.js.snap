// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component matches centered snapshot 1`] = `
<div
  className="container"
>
  <div
    className="wrapper"
  >
    <h6
      className="tag"
    >
      H6 Heading centered
    </h6>
  </div>
</div>
`;

exports[`Heading component matches className snapshot 1`] = `
<div
  className="container"
>
  <div
    className="wrapper"
  >
    <h4
      className="tag foobar"
    >
      H4 Heading with className
    </h4>
  </div>
</div>
`;

exports[`Heading component matches id snapshot 1`] = `
<div
  className="container"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="wrapper"
  >
    <h2
      className="tag"
      id="FooBaz"
    >
      H2 Heading with ID
    </h2>
  </div>
</div>
`;

exports[`Heading component matches uncentered snapshot 1`] = `
<div
  className="container"
>
  <div
    className="wrapper"
  >
    <h2
      className="tag"
    >
      H2 Heading
    </h2>
  </div>
</div>
`;
